index.js

```diff
@@ -113,14 +133,99 @@ module.exports = {
+        req.backstopReportUrl = self.cs.getUrl('/site-review/report/html_report/index.html');
+
+        var backstopPaths = {
+          "bitmaps_reference": self.apos.rootDir + "/data/backstop_data/bitmaps_reference",
+          "bitmaps_test": self.apos.rootDir + "/data/backstop_data/bitmaps_test",
+          "engine_scripts": self.apos.rootDir + "/data/backstop_data/engine_scripts",
+          "html_report": self.apos.rootDir + "/data/backstop_data/html_report",
+          "ci_report": self.apos.rootDir + "/data/backstop_data/ci_report"
+        };
+
+        backstopConfig.paths = backstopPaths;
+
+        backstopConfig.scenarios[0].url = req.deployToPath;
+        backstopConfig.scenarios[0].referenceUrl = req.deployFromPath;
+        return backstop('reference', {
+          config: backstopConfig
+        });
+      }).then(() => {
+        return backstop('test', { config: backstopConfig }).catch(() => {
+          // "Test failed," i.e. the report shows a difference. Not fatal
+        });
+      }).then(() => {
+        // This isn't very unique, but it's impractical to review reports in parallel anyway
+        return self.cs.syncFolder(self.apos.rootDir + "/data/backstop_data", "/site-review/report");
+      });
+    };
+
+    // Obtains a production URL for the given doc. Returns a
+    // promise. doc.workflowLocale must match the current locale.
+
+    self.getProductionUrl = function(req, doc) {
+      var deployToArray = self.getDeployToArrayForCurrentLocale(req);
+      return self.remoteApi('url', {
+        method: 'POST',
+        json: true,
+        body: {
+          workflowLocale: doc.workflowLocale,
+          workflowGuid: doc.workflowGuid
+        }
+      }, {
+        deployTo: deployToArray[0]
+      })
+      .then(function(result) {
+        if (result.status === 'ok') {
+          return result.url;
+        } else {
+          throw new Error('url remote api failed');
+        }
+      });
+    };
+
+    self.visualDiff = function(req) {
+      if (!self.isAdmin(req)) {
+        return '';
+      }
+      return self.partial('visualDiff',
+        {
+          workflowMode: req.session.workflowMode,
+          deployToPath: req.deployToPath,
+          deployFromPath: req.deployFromPath,
+          report: req.backstopReportUrl
+        }
+      );
+    };
+
     var superPageBeforeSend = self.pageBeforeSend;
     self.pageBeforeSend = function(req, callback) {
+      var review;
       if (!self.isAdmin(req)) {
         superPageBeforeSend(req);
         return callback(null);
       }
+
       return self.getActiveReview(req)
-      .then(function(review) {
+      .then(function(_review) {
+        review = _review;
+        // returns a promise and passes it to the next item in the chain.
+        // knows about req because it exists in the enclosure.
+
+        if (review && (req.data.page || req.data.piece)) {
+          return self.generateReport(req);
```

Comment:
We are pending on the entire review being generated before the response is sent for the page? How long might that take? If it's anything close to 20 seconds we have to look at an approach that doesn't risk a proxy server timeout.